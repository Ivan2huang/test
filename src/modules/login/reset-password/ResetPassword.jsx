import React, { useEffect, useState, useContext } from 'react';
import * as PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import ResetPasswordHeader from './ResetPasswordHeader';
import ResetPasswordForm from './ResetPasswordForm';
import VerifyAccountForm from './VerifyAccountForm';
import getLocale from '../../../i18n/getLocale';
import { IntlContext } from '../../../i18n/withIntlProvider';
import { isValidLanguageCode } from '../../../helpers/helpers';
import { getCookie } from '../../../helpers/auth';
import { LANG } from '../../../constants/auth';
import CONFIG from '../../../constants/config';

const ResetPassword = ({
  intl,
  resetPassword,
  router,
  validateResetPasswordRequest,
  isFirstTimeUser,
  dateOfBirthMatch,
  productName,
  setFormErrorIfDateOfBirthNoMatch,
  getProductName,
}) => {
  const {
    email,
    token,
    productName: productNameFromQuery,
    additionalVerification,
    lang,
    clientId,
  } = router.query;
  const [skipDOBValidation, setSkipDOBValidation] = useState(false);
  const setLocale = useContext(IntlContext);
  const selectedLang = lang || getCookie(LANG);
  const locale = isValidLanguageCode(selectedLang) ? selectedLang : getLocale();
  const selectedProductName = productName || productNameFromQuery;

  useEffect(() => {
    validateResetPasswordRequest(email, null, token, clientId);
    setLocale(locale);
  }, []);

  useEffect(() => {
    if (CONFIG.useProductNameFromCMS) {
      getProductName(intl.locale, productNameFromQuery || '');
    }
  }, [intl]);

  useEffect(() => {
    const skipDOBTimer = setTimeout(() => {
      const dateOfBirthValidationSuccess =
        dateOfBirthMatch && dateOfBirthMatch.errorState === false;
      setSkipDOBValidation(
        dateOfBirthValidationSuccess || additionalVerification === 'false',
      );
    });
    return () => {
      clearTimeout(skipDOBTimer);
    };
  }, [dateOfBirthMatch]);

  return skipDOBValidation ? (
    <>
      <ResetPasswordHeader
        userEmail={email}
        productName={selectedProductName}
        isFirstTimeUser={isFirstTimeUser}
      />
      <ResetPasswordForm
        resetPassword={resetPassword}
        email={email}
        token={token}
        isFirstTimeUser={isFirstTimeUser}
        productName={selectedProductName}
        clientId={clientId}
      />
    </>
  ) : (
    <VerifyAccountForm
      email={email}
      token={token}
      dateOfBirthMatch={dateOfBirthMatch}
      verifyDoB={validateResetPasswordRequest}
      setFormErrorIfDateOfBirthNoMatch={setFormErrorIfDateOfBirthNoMatch}
      clientId={clientId}
    />
  );
};
ResetPassword.propTypes = {
  intl: PropTypes.shape({}).isRequired,
  validateResetPasswordRequest: PropTypes.func.isRequired,
  resetPassword: PropTypes.func.isRequired,
  router: PropTypes.shape({}).isRequired,
  isFirstTimeUser: PropTypes.bool.isRequired,
  dateOfBirthMatch: PropTypes.shape({}),
  setFormErrorIfDateOfBirthNoMatch: PropTypes.func.isRequired,
  getProductName: PropTypes.func.isRequired,
  productName: PropTypes.string,
};
ResetPassword.defaultProps = {
  dateOfBirthMatch: {},
  productName: '',
};
export default injectIntl(ResetPassword);
