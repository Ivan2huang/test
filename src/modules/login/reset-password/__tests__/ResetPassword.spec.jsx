import React from 'react';
import { render, act } from '@testing-library/react';
import ResetPassword from '../ResetPassword';
import withIntl from '../../../../i18n/withIntlProvider';

jest.useFakeTimers();

jest.mock('../ResetPasswordHeader', () => props => (
  <div>
    <div>Header Component</div>
    <span>
      {Object.entries(props).map(([key, value]) => `${key} = ${value}`)}
    </span>
  </div>
));

jest.mock('../ResetPasswordForm', () => props => (
  <div>
    <span>Reset Password Form Component</span>
    <span data-id="props">
      {Object.entries(props).map(([key, value]) => `${key} = ${value}`)}
    </span>
  </div>
));

jest.mock('../VerifyAccountForm', () => props => (
  <div>
    <span>Verify Account Form Component</span>
    <span data-id="props">
      {Object.entries(props).map(([key, value]) => `${key} = ${value}`)}
    </span>
  </div>
));

describe('Reset Password Component', () => {
  const setUp = props => {
    const Component = withIntl(ResetPassword);
    return render(<Component {...props} />);
  };

  const props = {
    isFirstTimeUser: false,
    verifyRequired: 'verifyRequired',
    dateOfBirthMatch: { errorState: false },
    resetPassword: jest.fn(),
    setFormErrorIfDateOfBirthNoMatch: jest.fn(),
    validateResetPasswordRequest: jest.fn(),
    router: {
      query: {
        email: 'test@tests.com',
        token: '1k2j32',
        productName: 'product',
        clientId: 'clientA',
      },
    },
    getProductName: jest.fn(),
  };
  it('should validate email and token on load', () => {
    setUp(props);
    expect(props.validateResetPasswordRequest).toHaveBeenCalledTimes(1);
    expect(props.validateResetPasswordRequest).toHaveBeenCalledWith(
      'test@tests.com',
      null,
      '1k2j32',
      'clientA',
    );
  });

  it('should match snapshot when there is no dateOfBirthMatch error', () => {
    const newProps = {
      ...props,
      dateOfBirthMatch: { errorState: null },
    };
    const { container } = setUp(newProps);
    act(() => {
      jest.runAllTimers();
    });
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when there is no dateOfBirthMatch error is true', () => {
    const newProps = {
      ...props,
      dateOfBirthMatch: { errorState: true },
    };
    const { container } = setUp(newProps);
    act(() => {
      jest.runAllTimers();
    });
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when additionalVerification param is false and there is no dateOfBirthMatch error', () => {
    const newProps = {
      ...props,
      router: {
        query: {
          email: 'test@tests.com',
          token: '1k2j32',
          productName: 'product',
          additionalVerification: 'false',
          clientId: 'clientA',
        },
      },
      dateOfBirthMatch: { errorState: null },
    };
    const { container } = setUp(newProps);
    act(() => {
      jest.runAllTimers();
    });
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when additionalVerification param is false and dateOfBirthMatch is true', () => {
    const newProps = {
      ...props,
      router: {
        query: {
          email: 'test@tests.com',
          token: '1k2j32',
          productName: 'product',
          additionalVerification: 'false',
          clientId: 'clientA',
        },
      },
      dateOfBirthMatch: { errorState: true },
    };
    const { container } = setUp(newProps);
    act(() => {
      jest.runAllTimers();
    });
    expect(container).toMatchSnapshot();
  });
});
