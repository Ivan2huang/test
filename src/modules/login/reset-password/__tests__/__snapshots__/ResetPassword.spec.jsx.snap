// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset Password Component should match snapshot when additionalVerification param is false and dateOfBirthMatch is true 1`] = `
<div>
  <div>
    <div>
      Header Component
    </div>
    <span>
      userEmail = test@tests.com
      productName = product
      isFirstTimeUser = false
    </span>
  </div>
  <div>
    <span>
      Reset Password Form Component
    </span>
    <span
      data-id="props"
    >
      resetPassword = function mockConstructor() {
        return fn.apply(this, arguments);
      }
      email = test@tests.com
      token = 1k2j32
      isFirstTimeUser = false
      productName = product
      clientId = clientA
    </span>
  </div>
</div>
`;

exports[`Reset Password Component should match snapshot when additionalVerification param is false and there is no dateOfBirthMatch error 1`] = `
<div>
  <div>
    <div>
      Header Component
    </div>
    <span>
      userEmail = test@tests.com
      productName = product
      isFirstTimeUser = false
    </span>
  </div>
  <div>
    <span>
      Reset Password Form Component
    </span>
    <span
      data-id="props"
    >
      resetPassword = function mockConstructor() {
        return fn.apply(this, arguments);
      }
      email = test@tests.com
      token = 1k2j32
      isFirstTimeUser = false
      productName = product
      clientId = clientA
    </span>
  </div>
</div>
`;

exports[`Reset Password Component should match snapshot when there is no dateOfBirthMatch error 1`] = `
<div>
  <div>
    <span>
      Verify Account Form Component
    </span>
    <span
      data-id="props"
    >
      email = test@tests.com
      token = 1k2j32
      dateOfBirthMatch = [object Object]
      verifyDoB = function mockConstructor() {
        return fn.apply(this, arguments);
      }
      setFormErrorIfDateOfBirthNoMatch = function mockConstructor() {
        return fn.apply(this, arguments);
      }
      clientId = clientA
    </span>
  </div>
</div>
`;

exports[`Reset Password Component should match snapshot when there is no dateOfBirthMatch error is true 1`] = `
<div>
  <div>
    <span>
      Verify Account Form Component
    </span>
    <span
      data-id="props"
    >
      email = test@tests.com
      token = 1k2j32
      dateOfBirthMatch = [object Object]
      verifyDoB = function mockConstructor() {
        return fn.apply(this, arguments);
      }
      setFormErrorIfDateOfBirthNoMatch = function mockConstructor() {
        return fn.apply(this, arguments);
      }
      clientId = clientA
    </span>
  </div>
</div>
`;
